### PR TITLE
Time-over-time comparison - e2e tests

### LINKED ISSUE #43006: [Epic] Time-over-time comparison UI for aggregated columns
**Links**
- product doc: [Time-over-time comparison UI for aggregated columns](https://www.notion.so/metabase/Time-over-time-comparison-UI-for-aggregated-columns-032bb9b47d124731baa9c3c7f954727f)
- eng doc: [[Tech] Time-over-time comparison UI for aggregated columns](https://www.notion.so/metabase/Time-over-time-comparison-UI-for-aggregated-columns-d06c46884d924bb99b814ec9d8f9b114)
- feature branch: ~`time-over-time-comparison`~ PRs can go straight to `master`
- issue links:
    - #40313

### Milestone 1

```[tasklist]
- [ ] https://github.com/metabase/metabase/issues/43079
- [ ] https://github.com/metabase/metabase/issues/43070
- [ ] https://github.com/metabase/metabase/issues/43178
- [ ] https://github.com/metabase/metabase/issues/43071
- [ ] https://github.com/metabase/metabase/issues/43072
- [ ] https://github.com/metabase/metabase/issues/43073
```

### LINKED ISSUE #43073: [Testing plan] Time-over-time comparison
Testing plan for https://www.notion.so/metabase/Time-over-time-comparison-UI-for-aggregated-columns-032bb9b47d124731baa9c3c7f954727f

# Unit tests

- [x] For each of `offsetClause`, `diffOffsetClause`, `percentDiffOffsetClause` helpers:
  - [x] returns correct expression clause
  - [x] sets correct name
    - [x] singular `offset` (=-1)
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
    - [x] plural `offset` (<-1)
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
- [x] `AggregationPicker`
  - [x] does not display the option if there are no aggregations
  - [x] correct label
    - [x] single aggregation
    - [x] multiple aggregations
  - [x] calls `onAdd` when submitted
  - [x] does not call `onSelect` when submitted
- [x] `CompareAggregations`
  - [x] period input
    - [x] does not allow negative values
    - [x] does not allow non-integer values
  - [x] 1 aggregation
    - [x] does not display extra step
    - [x] cannot go back to extra step
  - [x] multiple aggregations
    - [x] displays extra step
    - [x] can go back to extra step
  - [x] submit button is disabled when no columns are selected
  - [x] submit button is disabled when offset value is empty
  - [x] calls `onSubmit`
    - [x] 1 column
    - [x] multiple columns

# e2e tests

- [x] Via aggregation picker in notebook editor
  - [x] is not displayed when there is no aggregation
  - [x] correct label
    - [x] 1 aggregation
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
    - [x] multiple aggregations
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
  - [x] adds new column(s)
    - [x] assert analytic event dispatched
    - [x] assert column name
    - [x] assert valid expression in the notebook editor widget
- [x] Via aggregation picker in chill mode (summarize)
  - [x] is not displayed when there is no aggregation
  - [x] correct label
    - [x] 1 aggregation
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
    - [x] multiple aggregations
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
  - [x] adds new column(s)
    - [x] assert analytic event dispatched
    - [x] assert column name
    - [x] assert valid expression in the notebook editor widget
- [x] Via plus button in table visualization
  - [x] is not displayed when there is no aggregation
  - [x] correct label
    - [x] 1 aggregation
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
    - [x] multiple aggregations
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
  - [x] adds new column(s)
    - [x] assert analytic event dispatched
    - [x] assert column name
    - [x] assert valid expression in the notebook editor widget
- [x] Via column header drill in table visualization
  - [x] is displayed only in columns created by aggregations
  - [x] is not displayed for data columns
  - [x] ~is not displayed for custom columns~
  - [x] is not displayed for breakout columns
    - [x] 1 aggregation
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
    - [x] multiple aggregations
      - [x] no breakout
      - [x] breakout on binned datetime column
      - [x] breakout on non-binned datetime column
      - [x] breakout on non-datetime column
  - [x] adds new column(s)
    - [x] assert analytic event dispatched
    - [x] assert column name
    - [x] assert valid expression in the notebook editor widget

### PR DESCRIPTION
**Pull request by kamilmielnik**
Closes #43073
Closes #43006

### How to verify

- PR implements e2e tests listed in #43073
- Assertions match [product expectations](https://www.notion.so/metabase/Time-over-time-comparison-UI-for-aggregated-columns-032bb9b47d124731baa9c3c7f954727f)
- CI is green

### CODE REVIEW

File: frontend/src/metabase/visualizations/click-actions/actions/CompareAggregationsAction/CompareAggregationsAction.tsx [hunk 1]
```diff
@@ -84,3 +86,19 @@ export const CompareAggregationsAction: LegacyDrill = ({
     },
   ];
 };
+
+export const getTitle = (
```
Comment (kamilmielnik) [inline frontend/src/metabase/visualizations/click-actions/actions/CompareAggregationsAction/CompareAggregationsAction.tsx:90]: Tests uncovered a bug in this component, here's a fix

File: e2e/test/scenarios/question/column-compare.cy.spec.ts [hunk 2]
```diff
@@ -10,134 +10,798 @@ import {
...
+        "Count (vs previous value)",
+        "Count (% vs previous value)",
+      ]);
     });
   });
 });
 
-function toggleColumnPicker() {
+function toggleColumnPickerItems(names: string[]) {
+  cy.findByTestId("column-picker").click({ force: true });
```
Comment (uladzimirdev) [inline e2e/test/scenarios/question/column-compare.cy.spec.ts:681]: nit. no need for `force`, same for the line after `for of`

```suggestion
  cy.findByTestId("column-picker").parent().click();
```

**Review by uladzimirdev — APPROVED** [review]

File: e2e/test/scenarios/question/column-compare.cy.spec.ts [hunk 3]
```diff
@@ -10,134 +10,798 @@ import {
...
+  }
+}
+
+function verifyBreakoutRequiredError() {
+  visualize();
+
+  cy.get("main")
+    .findByText("There was a problem with your question")
+    .should("exist");
+  cy.get("main").findByText("Show error details").should("exist").click();
```
Comment (uladzimirdev) [inline e2e/test/scenarios/question/column-compare.cy.spec.ts:801]: nit: both `findByText` and `click` make internal assertion and will fail if the element doesn't exist
```suggestion
  cy.get("main").findByText("Show error details").click();
```

File: e2e/test/scenarios/question/column-compare.cy.spec.ts [hunk 4]
```diff
@@ -10,134 +10,798 @@ import {
...
+    cy.findAllByTestId("header-cell").contains(name).should("exist");
+  }
+}
+
+function verifyBreakoutRequiredError() {
+  visualize();
+
+  cy.get("main")
+    .findByText("There was a problem with your question")
+    .should("exist");
```
Comment (uladzimirdev) [inline e2e/test/scenarios/question/column-compare.cy.spec.ts:800]: same, `findByText` will fail, if the element doesn't exist, maybe `be.visible` is what I want to check?

File: e2e/test/scenarios/question/column-compare.cy.spec.ts (hunk 3 above)
Comment (kamilmielnik) [reply e2e/test/scenarios/question/column-compare.cy.spec.ts]: Usually I want to be explicit that it's an intentional assertion and not a part of "test flow", but yeah, it does not make much sense here: 836e13e7af111c70ef1b1f6e224634815c342cfd

File: e2e/test/scenarios/question/column-compare.cy.spec.ts (hunk 4 above)
Comment (kamilmielnik) [reply e2e/test/scenarios/question/column-compare.cy.spec.ts]: I just wanted to have an explicit assertion.
`be.visible` is a better assertion though - updated everywhere: 59ea9e444219174a03d82fb7345450986fec361

**Pull request merged**